Sources/MultiSets/CountedSet/CountedSet+Collection.swift

```diff
@@ -0,0 +1,105 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift Collections open source project
+//
+// Copyright (c) 2021 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+//
+//===----------------------------------------------------------------------===//
+
+extension CountedSet: Collection {
+  /// The position of an element in a counted set.
+  @frozen
+  public struct Index: Comparable {
+    /// An index in the underlying dictionary storage of a counted set.
+    ///
+    /// This is used to distinguish between indices that point to elements with
+    /// different values.
+    @usableFromInline
+    let storageIndex: Dictionary<Element, UInt>.Index
+
+    /// The relative position of the element.
+    ///
+    /// This doesn't actually correspond to a distinct part of memory. Instead,
+    /// it is used to distinguish between indices that point to elements of the
+    /// same value.
+    ///
+    /// For example, the first index pointing to a value has an index of 0, the
+    /// second index pointing to that value will have an index of 1, and so on.
+    ///
+    /// When a counted set is subscripted with an index, the stored multiplicity
+    /// is retrieved using the `storageIndex` and compared with the `position`
+    /// to determine the index's validity.
+    @usableFromInline
+    let position: UInt
+
+    @inlinable
+    public static func < (
+      lhs: CountedSet<Element>.Index,
+      rhs: CountedSet<Element>.Index
+    ) -> Bool {
+      guard lhs.storageIndex != rhs.storageIndex else {
+        return lhs.storageIndex < rhs.storageIndex
+      }
+      return lhs.position < rhs.position
+    }
+
+    @usableFromInline
+    init(storageIndex: Dictionary<Element, UInt>.Index, position: UInt) {
+      self.storageIndex = storageIndex
+      self.position = position
+    }
+  }
+
+  @inlinable
+  public func index(after i: Index) -> Index {
+    guard i.position + 1 < rawValue[i.storageIndex].value else {
+      return Index(
+        storageIndex: rawValue.index(after: i.storageIndex),
+        position: 0
+      )
+    }
+
+    return Index(storageIndex: i.storageIndex, position: i.position + 1)
+  }
+
+  @inlinable
+  public subscript(position: Index) -> Element {
+    let keyPair = rawValue[position.storageIndex]
+    precondition(
+      position.position < keyPair.value,
+      "Attempting to access CountedSet elements using an invalid index"
+    )
+    return keyPair.key
+  }
+
+  ///  The number of elements in the set.
+  ///
+  /// - Complexity: O(*k*), where *k* is the number of unique elements in the
+  /// set.
+  @inlinable
+  public var count: Int {
+    Int(rawValue.values.reduce(.zero, +))
```

Comment:
Storing the count would mean this is no longer an implicit data structure. I’m not really sure that’s a worthwhile tradeoff.

Is there precedent for doing that with similar data structures?

Reply:
Oh definitely -- `Set` and `Dictionary` do it, for example, to prevent having to scan their bitmap to figure out their element count.

I don't think having to store one extra integer matters much in this case -- the hash table in the underlying dictionary already comes with way too much memory overhead to call this a truly implicit data structure. 🙈

Reply:
I was thinking it was an implicit data structure _for a dictionary_, but okay then.